### ISSUE TITLE
Add a recent checkpoint

### ISSUE THREAD

**Issue by daira** [zcash/zcash#6224]
\@miodrag on the Community Discord pointed out that there have been no new checkpoints since block 525000 (1st of May 2019):
> Absence of fresh checkpoints makes the node unable to skip expensive checks during sync. Furthermore, it renders `-ibdskiptxverification` option useless.

**Comment by sandakersmann** [comment]
@daira I can do this if you want.

**Comment by miodragpop** [comment]
Allow me to elaborate here why this is a necessary, but not quite sufficient condition for relaxed verification during synchronization. Zcash (unlike Bitcoin) does not support "headers-first" download mode. Block headers are read ahead of the blocks, but as long as the headers tip doesn't reach next checkpoint, zcashd is unaware of it's existence. Thus zcashd falls back to expensive block validations more often than it should, and it hurts sync performance. You can raise the probability of skipping those expensive checks by placing many checkpoints in the code, with narrow gap between two consecutive checkpoints. Ideally (imho), "headers-first" mode could be implemented; only the last checkpoint would be the significant one then ... or "assume-valid-hash" as it's replacement.

**Comment by sandakersmann** [comment]
So add a checkpoint for each 100k block? Is that what you are advocating for?

**Comment by miodragpop** [comment]
> So add a checkpoint for each 100k block? Is that what you are advocating for?

Let's say that would be plausible ad hoc workaround in order to shorten sync-from-scratch time, but personally I would prefer headers-first implementation.

**Comment by sandakersmann** [comment]
From what I can see headers-first is implemented in Zcash:

https://github.com/zcash/zcash/commit/341735eb8f42e898cf9d4d130709471e5d01abe2

**Comment by miodragpop** [comment]
> From what I can see headers-first is implemented in Zcash:
> 
> [341735e](https://github.com/zcash/zcash/commit/341735eb8f42e898cf9d4d130709471e5d01abe2)

Sure, but it's somehow broken. If you try syncing full node, you'll see that blocks are being downloaded in parallel with headers, while headers tip is just a few thousands blocks ahead of blocks tip and not even close to the current network chain tip.

![Screenshot from 2022-10-30 19-01-57](https://user-images.githubusercontent.com/37712022/198899494-c2440392-4b30-4c29-a1aa-9b7d1a57f830.png)

**Comment by miodragpop** [comment]
I'll try to investigate this, it deserves a separate issue filed.

**Comment by sandakersmann** [comment]
Sounds like standard behavior to me

**Comment by miodragpop** [comment]
> Sounds like standard behavior to me

bitcoin: synchronizes ALL missing headers before it starts downloading full blocks, thus knows if blocks are ancestors of the last checkpoint. 

zcash: synchronizes SOME missing headers before it starts downloading full blocks, doesn't know if incoming blocks are ancestors of the last checkpoint.

Below is the console output screenshot of btc node syncing, at the moment when it switches from headers to blocks....
![btcsync](https://user-images.githubusercontent.com/37712022/198952190-d3ee5ec7-9613-43b1-bbee-c560157bca5b.png)

**Comment by sandakersmann** [comment]
Isn't the problem just that you start a client with no checkpoint after the blocks already synced?

**Comment by miodragpop** [comment]
No, that's only one small fraction of a problem. Even with checkpoints added at each 100k blocks zcashd would spend more time in expensive checks than in relaxed checks, because of headers tip advancing slowly ahead of blocks tip. As I've said, frequent checkpoints would just raise chances to spend more time in relaxed checks mode, better than as is but actually a workaround. Real solution would be proper headers-first mode when we would be able to deterministically claim: "Expensive checks will be skipped for all historical blocks which are ancestors of the last known checkpoint". We can't say that now, for zcashd.

**Comment by miodragpop** [comment]
A checkpoint is considered known and qualified only if it's header is downloaded. Prior to that, it's nonexistent in block index map.

**Comment by miodragpop** [comment]
I've found the cause of premature block download start. If PR #6231 gets merged, then adding of multiple checkpoints as a workaround won't be necessary (only the last checkpoint and matching minimum chainwork will be relevant).

**Comment by str4d** [comment]
Closed by #6228.
